core/scheduler/resolver/lineage_resolver.go

```diff
@@ -168,8 +168,19 @@ type upstreamCandidate struct {
 }
 
 func (r *LineageResolver) pruneLineage(lineage *scheduler.JobLineageSummary, maxUpstreamsPerLevel, depth int) *scheduler.JobLineageSummary {
-	// base case: stop if max depth reached or number of upstreams is already within limit
-	if depth > MaxLineageDepth || len(lineage.Upstreams) <= maxUpstreamsPerLevel {
+	if depth > MaxLineageDepth {
```

Comment:
can we avoid to use function recursive? use stack data type instead

function recursive uses `stack` frame allocation, which is limited, and can lead to stack overflow. meanwhile using stack data type is more safe, as it uses `heap` memory